### PR TITLE
Update monailabel version and other improvements to fasten training process

### PR DESCRIPTION
**Pull request by SachidanandAlle**
- Fixes for `blind_spot_patch` 
  - support both tensor and numpy version
  - support (patches x h x w x channel) or (h x w x channel) to avoid non-required channel addition
- Remove unwanted code
  - Remove handlers.py (unused)
  - Remove non relevant transforms
  - Unused imports (sorted)
- Fix `BlindSpotPatch` Transform (make it randomizable/non-cacheable)
- Upgrade monailabel version
  - Fixes to disable meta tracking
  - Avoid using Deprecated interface for InferTask
- Stub code in main.py to run/test train function locally...

Signed-off-by: Sachidanand Alle <sachidanand.alle@gmail.com>

### CODE REVIEW

File: dataprocessing/dataloaders.py [hunk 1]
```diff
@@ -331,14 +331,20 @@ def __call__(self, data):
...
     num_sample = int(height * width * (1 - ratio))
 
-    mask = np.ones(input.shape)
-    output = np.array(input)
+    if isinstance(input, torch.Tensor):
+        mask = torch.ones_like(input)
+        output = input.clone().detach()
+    else:
+        mask = np.ones(input.shape)
+        output = np.array(input)
```
Comment (gunjan-sh) [inline dataprocessing/dataloaders.py:347]: nice

File: impartial/api/lib/transforms.py [hunk 2]
```diff
@@ -109,25 +84,27 @@ def __call__(self, data):
...
+            input, mask = blind_spot_patch(d[key], ratio=self.ratio, size_window=self.size_window)
 
-        d["input"] = np.moveaxis(d["input"], -1, 0)
-        d["mask"] = np.moveaxis(d["mask"], -1, 0)
+            input = torch.moveaxis(input, -1, 0) if isinstance(input, torch.Tensor) else np.moveaxis(input, -1, 0)
+            mask = torch.moveaxis(mask, -1, 0) if isinstance(mask, torch.Tensor) else np.moveaxis(mask, -1, 0)
 
+            d[self.input] = input
+            d[self.mask] = mask
         return d
```
Comment (gunjan-sh) [inline impartial/api/lib/transforms.py:108]: Didn't understand the need of a for loop here, since final update to   d[self.input] & d[self.mask] in the loop will be returned ? I am thinking this:


    def __call__(self, data):
        d = dict(data)
        input, mask = blind_spot_patch(d['image'], ratio=self.ratio, size_window=self.size_window)

        input = torch.moveaxis(input, -1, 0) if isinstance(input, torch.Tensor) else np.moveaxis(input, -1, 0)
        mask = torch.moveaxis(mask, -1, 0) if isinstance(mask, torch.Tensor) else np.moveaxis(mask, -1, 0)

        d[self.input] = input
        d[self.mask] = mask

        return d

File: impartial/api/lib/transforms.py [hunk 3]
```diff
@@ -176,214 +153,3 @@ def __call__(self, data):
...
-        tx, ty = location[0], location[1]
-
-        pos = d.get(self.foreground)
-        pos = (np.array(pos) - (tx, ty)).astype(int).tolist() if pos else []
-
-        neg = d.get(self.background)
-        neg = (np.array(neg) - (tx, ty)).astype(int).tolist() if neg else []
-
-        d[self.guidance] = [pos, neg]
-        return d
```
Comment (gunjan-sh) [inline impartial/api/lib/transforms.py:389]: @ricdodds since you added these, please confirm if these can be removed ?

File: impartial/api/main.py [hunk 4]
```diff
@@ -111,3 +109,40 @@ def init_trainers(self) -> Dict[str, TrainTask]:
...
+            "max_epochs": 100,
+            # "dataset": "CacheDataset",
+        },
+    )
+
+
+if __name__ == "__main__":
+    # export PYTHONPATH=ImPartial/impartial:ImPartial/impartial/api
+    # python main.py
+    main()
```
Comment (gunjan-sh) [inline impartial/api/main.py:148]: thanks, this is useful

**Review by gunjan-sh — COMMENTED** [review]
- did some benchmarking on the changes.
- don't see an effect of using a cached dataset?
- see summary table bellow, timings are for single epoch :
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:"Segoe UI";font-size:12.0pt'>


<p style='margin:0in;font-family:"Segoe UI";font-size:12.0pt'>Dataset :
Vectra2CH, Epoch =100, Patches : 4096, Unet_base : 64 </p>

<div style='direction:ltr'>



Batch size | 128 | 128 | 8 | 8
:-- | :-- | :-- | :-- | :--
Dataset type | Nocache | Cache | Nocache | Cache
SupervisedTrainer   (sec) | 24.254 | 24.329 | 29.061 | 29.393
SupervisedEvaluator   (sec) | 00.116 | 00.157 | 00.738 | 00.747



</div>


</body>

</html>

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:"Segoe UI";font-size:12.0pt'>


<p style='margin:0in;font-family:"Segoe UI";font-size:12.0pt'>Training on AWS
EC2 g4dn.2xlarge Single GPU instance </p>

<div style='direction:ltr'>



Instance Size | GPU | vCPUs | Memory (GiB) | Instance Storage (GB)
-- | -- | -- | -- | --
g4dn.2xlarge | 1 x NVIDIA T4 | 8 | 32 | 1 x 225 NVMe SSD



</div>


</body>

</html>

File: impartial/api/lib/transforms.py (hunk 2 above)
Comment (SachidanandAlle) [reply impartial/api/lib/transforms.py:108]: I agree.. this is some stupid maptransform (key) based interface.. u supply single/multiple key to consume/produce something.. in this case.. its one key.. so look doesn't matter..

**Comment by SachidanandAlle** [comment]
> * did some benchmarking on the changes.
> * don't see an effect of using a cached dataset?
> * see summary table bellow, timings are for single epoch :
> 
> Dataset : Vectra2CH, Epoch =100, Patches : 4096, Unet_base : 64
> 
> Batch size	128	128	8	8
> Dataset type	Nocatche	Catche	Nocatche	Catche
> SupervisedTrainer (sec)	24.254	24.329	29.061	29.393
> SupervisedEvaluator (sec)	00.116	00.157	00.738	00.747
> Training on AWS EC2 g4dn.2xlarge Single GPU instance
> 
> Instance Size	GPU	vCPUs	Memory (GiB)	Instance Storage (GB)
> g4dn.2xlarge	1 x NVIDIA T4	8	32	1 x 225 NVMe SSD

1. Have you upgraded monailabel?
2. Important change is disable tracking.. you get 2x boost on training time.. you can enable and disable and see if there is any difference at the end..
3. what was the previous numbers without these changes - 25 sec?  I had same thing on myside.. so after all these changes mentioned in the PR.. i could bring it down to 7 sec (3x times faster)
4. cachedataset vs smartcacheddataset (which is the default) is almost same.. I tried to fix some of the transform so that they are really random and don't get cached wrt functionality

**Comment by SachidanandAlle** [comment]
If you are not sure about if disable tracking is disabled.. you can also add by yourself..

https://github.com/nadeemlab/ImPartial/blob/main/impartial/api/lib/trainers/impartial.py#L165

```python
from monai.data import
set_track_meta(False)
``` 
explicitly in your app code.

But this won't be essential if you are using 0.6.rc2 or above version of monailabel..

**Comment by gunjan-sh** [comment]
> > * did some benchmarking on the changes.
> > * don't see an effect of using a cached dataset?
> > * see summary table bellow, timings are for single epoch :
> > 
> > Dataset : Vectra2CH, Epoch =100, Patches : 4096, Unet_base : 64
> > Batch size	128	128	8	8
> > Dataset type	Nocatche	Catche	Nocatche	Catche
> > SupervisedTrainer (sec)	24.254	24.329	29.061	29.393
> > SupervisedEvaluator (sec)	00.116	00.157	00.738	00.747
> > Training on AWS EC2 g4dn.2xlarge Single GPU instance
> > Instance Size	GPU	vCPUs	Memory (GiB)	Instance Storage (GB)
> > g4dn.2xlarge	1 x NVIDIA T4	8	32	1 x 225 NVMe SSD
> 
> 1. Have you upgraded monailabel?
> 2. Important change is disable tracking.. you get 2x boost on training time.. you can enable and disable and see if there is any difference at the end..
> 3. what was the previous numbers without these changes - 25 sec?  I had same thing on myside.. so after all these changes mentioned in the PR.. i could bring it down to 7 sec (3x times faster)
> 4. cachedataset vs smartcacheddataset (which is the default) is almost same.. I tried to fix some of the transform so that they are really random and don't get cached wrt functionality


1. Yes, all the times above are on updated Monailabel :
    monai==1.0.1
    monailabel==0.6.0rc2
 
2. Current "main" branch without your changes and updated monailabel w/ smartdataset is around 34 sec (per epoch). With your PR changes the train time is around 25 secs (speed up of 9 sec?)

3. I manually disabled the tracking setting "set_track_meta(False)"  inside _create_trainer function. But, I don't see any improvement on my end with that change. Still seeing around 25 sec per epoch w/ and wo/ the flag set.

**Comment by SachidanandAlle** [comment]
if you are using 0.6.0rc2 version, tracking is already disabled.
how about you use single gpu? epoch per sec is still same?
and if possible, try on another GPU of different type (even single gpu is fine)

**Comment by gunjan-sh** [comment]
Yes, been using 0.6.0rc2 version and also earlier reported times were for AWS instances with 1GPU. 

Also, per you suggestion, adding more benchmarking on better servers, but our target is still to run on commodity AWS GPU instance. 

- AWS GPU instance (A10G)

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:"Segoe UI";font-size:12.0pt'>


<div style='direction:ltr;border-width:100%'>

<div style='direction:ltr;margin-top:0in;margin-left:0in;width:7.109in'>

<div style='direction:ltr;margin-top:0in;margin-left:0in;width:7.109in'>

<div style='direction:ltr'>



Batch size | 128 | 128 | 8 | 8
:-- | :-- | :-- | :-- | :--
Dataset type | SmartCache | Cache | SmartCache | Cache
SupervisedTrainer   (sec) | 12.231 | 12.288 | 18.599 | 16.128
SupervisedEvaluator   (sec) | 02.235 | 02.056 | 04.343 | 01.987



</div>

</div>

</div>

</div>


</body>

</html>


- on-premise mskcc server using 1 GPU (A40) 

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:"Segoe UI";font-size:12.0pt'>


<div style='direction:ltr;border-width:100%'>

<div style='direction:ltr;margin-top:0in;margin-left:0in;width:6.9604in'>

<div style='direction:ltr;margin-top:0in;margin-left:0in;width:6.9604in'>

<div style='direction:ltr'>



Batch size | 128 | 128 | 8 | 8
:-- | :-- | :-- | :-- | :--
Dataset type | SmartCache | Cache | SmartCache | Cache
SupervisedTrainer   (sec) | 10.899 | 10.483 | 18.181 | 16.345
SupervisedEvaluator   (sec) | 02.981 | 02.505 | 04.343 | 02.498



</div>

</div>

</div>

</div>


</body>

</html>
